File: examples/nirfmx/lte/csharp/VS2010/RFmxLteULPvtSingleCarrier/cs/RFmxLteULPvtSingleCarrier.cs
/*Steps:
1. Open a new RFmx Session.
2. Configure Frequency Reference.
3. Configure basic signal properties (Center Frequency, Reference Level and External Attenuation).
4. Configure Trigger Type and Trigger Parameters.
5. Configure Carrier Bandwidth.
6. Select PvT measurement and enable Traces.
7. Configure Duplex Scheme.
8. Configure Measurement Methods.
9. Configure Averaging Parameters for PvT measurement.
10. Initiate the Measurement.
11. Fetch PvT  Traces and Measurements.
12. Close RFmx Session.*/

using System;
using NationalInstruments.RFmx.InstrMX;
using NationalInstruments.RFmx.LteMX;

namespace NationalInstruments.Examples.RFmxLteULPvtSingleCarrier
{
    public class RFmxLteULPvtSingleCarrier
    {
        RFmxInstrMX instrSession;
        RFmxLteMX lte;

        string resourceName = "RFSA";
        string frequencyReferenceSource = RFmxInstrMXConstants.OnboardClock;
        string iqPowerEdgeTriggerSource = "0";

        double frequencyReferenceFrequency = 10e6;                                             /* Hz */
        double centerFrequency = 1.95e+9;                                                      /* Hz */
        double referenceLevel = 0.0;                                                           /* dBm */
        double externalAttenuation = 0.0;                                                      /* dB */
        double iqPowerEdgeTriggerLevel = -20.0;                                                /* dB */
        double triggerDelay = 0.0;                                                             /* seconds */
        double minimumQuietTimeDuration = 50.0e-6;								               /* seconds */
        double timeout = 10.0;                                                                /* seconds */
        double componentCarrierBandwidth = 10e6;                                               /* Hz */
        double componentCarrierFrequency = 0.0;                                                /* Hz */
        bool enableTrigger = true;
        int averagingCount = 10;
        int cellID = 0;

        double offPowerExclusionBefore = 0.0, offPowerExclusionAfter = 0.0;

        RFmxLteMXUplinkDownlinkConfiguration uplinkDownlinkConfiguration = RFmxLteMXUplinkDownlinkConfiguration.Configuration0;
        RFmxLteMXPvtMeasurementMethod measurementMethod = RFmxLteMXPvtMeasurementMethod.Normal;
        RFmxLteMXPvtAveragingEnabled averagingEnabled = RFmxLteMXPvtAveragingEnabled.False;
        RFmxLteMXPvtAveragingType averagingType = RFmxLteMXPvtAveragingType.Rms;
        RFmxLteMXTriggerMinimumQuietTimeMode minimumQuietTimeMode = RFmxLteMXTriggerMinimumQuietTimeMode.Auto;
        RFmxLteMXIQPowerEdgeTriggerLevelType iqPowerEdgeTriggerLevelType = RFmxLteMXIQPowerEdgeTriggerLevelType.Relative;
        RFmxLteMXIQPowerEdgeTriggerSlope iqPowerEdgeTriggerSlope = RFmxLteMXIQPowerEdgeTriggerSlope.Rising;
        RFmxLteMXDuplexScheme duplexScheme = RFmxLteMXDuplexScheme.Tdd;

        RFmxLteMXPvtMeasurementStatus measurementStatus;
        double meanAbsoluteOFFPowerBefore, meanAbsoluteOFFPowerAfter, meanAbsoluteONPower, burstWidth;
        AnalogWaveform<float> signalPower, absoluteLimit;

        public void Run()
        {
            try
            {
                InitializeInstr();
                ConfigureLte();
                RetrieveResults();
                PrintResults();
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
            finally
            {
                /* Close session */
                CloseSession();
                Console.WriteLine("Press any key to exit");
                Console.ReadKey();
            }
        }


        private void InitializeInstr()
        {
            /* Create a new RFmx Session */
            instrSession = new RFmxInstrMX(resourceName, "");
        }

        private void ConfigureLte()
        {
            /* Get Lte signal */
            lte = instrSession.GetLteSignalConfiguration();

            /* Configure measurement */
            instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequencyReferenceFrequency);

            lte.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation);

            lte.ConfigureIQPowerEdgeTrigger("", iqPowerEdgeTriggerSource, iqPowerEdgeTriggerSlope, iqPowerEdgeTriggerLevel,
                triggerDelay, minimumQuietTimeMode, minimumQuietTimeDuration, iqPowerEdgeTriggerLevelType, enableTrigger);

            lte.ComponentCarrier.Configure("", componentCarrierBandwidth,
                                                componentCarrierFrequency, cellID);

            lte.SelectMeasurements("", RFmxLteMXMeasurementTypes.Pvt, true);

            lte.ConfigureDuplexScheme("", duplexScheme, uplinkDownlinkConfiguration);

            lte.Pvt.Configuration.ConfigureMeasurementMethod("", measurementMethod);

            lte.Pvt.Configuration.ConfigureOffPowerExclusionPeriods("", offPowerExclusionBefore, offPowerExclusionAfter);

            lte.Pvt.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType);

            lte.Initiate("", "");

        }

        private void RetrieveResults()
        {
            /* Retrieve results */

            lte.Pvt.Results.FetchSignalPowerTrace("", timeout,
                    ref signalPower, ref absoluteLimit);

            lte.Pvt.Results.FetchMeasurement("", timeout, out measurementStatus, out meanAbsoluteOFFPowerBefore,
                                                         out meanAbsoluteOFFPowerAfter, out meanAbsoluteONPower,
                                                         out burstWidth);
        }

        private void PrintResults()
        {
            Console.WriteLine("\n********** Measurement **********");

            Console.WriteLine("Status                               : {0}", measurementStatus);
            Console.WriteLine("Mean Absolute OFF Power Before (dBm) : {0}", meanAbsoluteOFFPowerBefore);
            Console.WriteLine("Mean Absolute OFF Power After (dBm)  : {0}", meanAbsoluteOFFPowerAfter);
            Console.WriteLine("Mean Absolute ON Power (dBm)         : {0}", meanAbsoluteONPower);
            Console.WriteLine("Burst Width (s)                      : {0}", burstWidth);
        }

        private void CloseSession()
        {
            try
            {
                if (lte != null)
                {
                    lte.Dispose();
                    lte = null;
                }

                if (instrSession != null)
                {
                    instrSession.Close();
                    instrSession = null;
                }
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
        }

        static private void DisplayError(Exception ex)
        {
            Console.WriteLine("ERROR:\n" + ex.GetType() + ": " + ex.Message);
        }
    }
}

File: examples/nirfmx/lte/csharp/VS2010/RFmxLteULPvtSingleCarrier/vb/RFmxLteULPvtSingleCarrier.vb
'Steps:
'1. Open a new RFmx Session.
'2. Configure Frequency Reference.
'3. Configure basic signal properties (Center Frequency, Reference Level and External Attenuation).
'4. Configure Trigger Type and Trigger Parameters.
'5. Configure Carrier Bandwidth.
'6. Select PvT measurement and enable Traces.
'7. Configure Duplex Scheme.
'8. Configure Measurement Methods.
'9. Configure Averaging Parameters for PvT measurement.
'10. Initiate the Measurement.
'11. Fetch PvT  Traces and Measurements.
'12. Close RFmx Session.


Imports NationalInstruments.RFmx.InstrMX
Imports NationalInstruments.RFmx.LteMX

Public Class RFmxLteULPvtSingleCarrier
    Private instrSession As RFmxInstrMX
    Private lte As RFmxLteMX

    Private resourceName As String = "RFSA"
    Private frequencyReferenceSource As String = RFmxInstrMXConstants.OnboardClock
    Private iqPowerEdgeTriggerSource As String = "0"

    Private frequencyReferenceFrequency As Double = 10000000.0         'Hz 
    Private centerFrequency As Double = 1950000000.0                   'Hz 
    Private referenceLevel As Double = 0.0                             'dBm
    Private externalAttenuation As Double = 0.0                        'dB
    Private iqPowerEdgeTriggerLevel As Double = -20.0                  'dB
    Private triggerDelay As Double = 0.0                               'seconds
    Private minimumQuietTimeDuration As Double = 0.00005               'seconds
    Private timeout As Double = 10.0                                   'seconds
    Private componentCarrierBandwidth As Double = 10000000.0           'Hz
    Private componentCarrierFrequency As Double = 0.0                  'Hz
    Private enableTrigger As Boolean = True
    Private averagingCount As Integer = 10
    Private cellID As Integer = 0

    Private offPowerExclusionBefore As Double = 0.0                    'seconds
    Private offPowerExclusionAfter As Double = 0.0                     'seconds 

    Private uplinkDownlinkConfiguration As RFmxLteMXUplinkDownlinkConfiguration = RFmxLteMXUplinkDownlinkConfiguration.Configuration0
    Private measurementMethod As RFmxLteMXPvtMeasurementMethod = RFmxLteMXPvtMeasurementMethod.Normal
    Private averagingEnabled As RFmxLteMXPvtAveragingEnabled = RFmxLteMXPvtAveragingEnabled.[False]
    Private averagingType As RFmxLteMXPvtAveragingType = RFmxLteMXPvtAveragingType.Rms
    Private minimumQuietTimeMode As RFmxLteMXTriggerMinimumQuietTimeMode = RFmxLteMXTriggerMinimumQuietTimeMode.Auto
    Private iqPowerEdgeTriggerLevelType As RFmxLteMXIQPowerEdgeTriggerLevelType = RFmxLteMXIQPowerEdgeTriggerLevelType.Relative
    Private iqPowerEdgeTriggerSlope As RFmxLteMXIQPowerEdgeTriggerSlope = RFmxLteMXIQPowerEdgeTriggerSlope.Rising
    Private duplexScheme As RFmxLteMXDuplexScheme = RFmxLteMXDuplexScheme.Tdd

    Private measurementStatus As RFmxLteMXPvtMeasurementStatus
    Private meanAbsoluteOFFPowerBefore As Double, meanAbsoluteOFFPowerAfter As Double, meanAbsoluteONPower As Double, burstWidth As Double
    Private signalPower As AnalogWaveform(Of Single), absoluteLimit As AnalogWaveform(Of Single)

    Public Sub Run()
        Try
            InitializeInstr()
            ConfigureLte()
            RetrieveResults()
            PrintResults()
        Catch ex As Exception
            DisplayError(ex)
        Finally
            ' Close session 

            CloseSession()
            Console.WriteLine("Press any key to exit")
            Console.ReadKey()
        End Try
    End Sub


    Private Sub InitializeInstr()
        ' Create a new RFmx Session 

        instrSession = New RFmxInstrMX(resourceName, "")
    End Sub

    Private Sub ConfigureLte()
        ' Get Lte signal 

        lte = instrSession.GetLteSignalConfiguration()

        ' Configure measurement 

        instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequencyReferenceFrequency)

        lte.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation)

        lte.ConfigureIQPowerEdgeTrigger("", iqPowerEdgeTriggerSource, iqPowerEdgeTriggerSlope, iqPowerEdgeTriggerLevel, triggerDelay,
                                        minimumQuietTimeMode, minimumQuietTimeDuration, iqPowerEdgeTriggerLevelType, enableTrigger)

        lte.ComponentCarrier.Configure("", componentCarrierBandwidth, componentCarrierFrequency, cellID)

        lte.SelectMeasurements("", RFmxLteMXMeasurementTypes.Pvt, True)

        lte.ConfigureDuplexScheme("", duplexScheme, uplinkDownlinkConfiguration)

        lte.Pvt.Configuration.ConfigureMeasurementMethod("", measurementMethod)

        lte.Pvt.Configuration.ConfigureOffPowerExclusionPeriods("", offPowerExclusionBefore, offPowerExclusionAfter)

        lte.Pvt.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType)

        lte.Initiate("", "")

    End Sub

    Private Sub RetrieveResults()
        ' Retrieve results 


        lte.Pvt.Results.FetchSignalPowerTrace("", timeout, signalPower, absoluteLimit)

        lte.Pvt.Results.FetchMeasurement("", timeout, measurementStatus, meanAbsoluteOFFPowerBefore, meanAbsoluteOFFPowerAfter,
                                         meanAbsoluteONPower, burstWidth)
    End Sub

    Private Sub PrintResults()
        Console.WriteLine(vbLf & "********** Measurement **********")

        Console.WriteLine("Status                               : {0}", measurementStatus)
        Console.WriteLine("Mean Absolute OFF Power Before (dBm) : {0}", meanAbsoluteOFFPowerBefore)
        Console.WriteLine("Mean Absolute OFF Power After (dBm)  : {0}", meanAbsoluteOFFPowerAfter)
        Console.WriteLine("Mean Absolute ON Power (dBm)         : {0}", meanAbsoluteONPower)
        Console.WriteLine("Burst Width (s)                      : {0}", burstWidth)
    End Sub

    Private Sub CloseSession()
        Try
            If lte IsNot Nothing Then
                lte.Dispose()
                lte = Nothing
            End If

            If instrSession IsNot Nothing Then
                instrSession.Close()
                instrSession = Nothing
            End If
        Catch ex As Exception
            DisplayError(ex)
        End Try
    End Sub

    Private Shared Sub DisplayError(ex As Exception)
        Console.WriteLine("ERROR:" & vbLf & Convert.ToString(ex.[GetType]()) & ": " & ex.Message)
    End Sub
End Class
